src/logger/logger.ts

```diff
@@ -1,12 +1,100 @@
-import pino from "pino"
+import _ from "lodash"
+import pino, { Logger as PinoLogger } from "pino"
+
+import { StackFrame, get, parse } from "./stack-trace"
+
+interface WarnMeta {
+  params: Record<string, unknown>
+}
+
+interface ErrorMeta {
+  error: unknown
+}
+
+const MAX_STACK_DEPTH = 3
+
+export interface TraceMeta {
+  file: string
+  line: number
+  func: string
+}
+
+export const getTraceMeta = (frames: StackFrame[]): TraceMeta[] =>
+  frames.map((frame) => ({
+    file: frame.getFileName(),
+    line: frame.getLineNumber(),
+    func: frame.getFunctionName(),
+  }))
+
+export class Logger {
+  _logger: PinoLogger
+
+  constructor(logger: PinoLogger) {
+    this._logger = logger
+  }
+
+  public child = (
+    bindings: Record<string, unknown> & { module: string }
+  ): Logger => {
+    const child = this._logger.child(bindings)
+    return new Logger(child)
+  }
+
+  public error = (
+    message: string,
+    meta: WarnMeta & ErrorMeta & Record<string, unknown>
+  ): void => {
+    const stackFrames = meta.error instanceof Error ? parse(meta.error) : get()
+    const trace = getTraceMeta(stackFrames.slice(0, MAX_STACK_DEPTH))
+
+    this._logger.error(message, {
+      isomer: {
```

Comment:
i'm not too opinionated wrt `isomer` key tbh, it's just that in some logs i've observed, the log metadata is grouped by the origin (`aws/dd` here) 

wrt `error/err` it's there already, as part of `meta` - `type meta = WarnMeta & ErrorMeta & Record<string, unknown>`

and `ErrorMeta` is `{ error: unknown }`. i chose `unknown` over `error` because in `try/catch` loops, the `error` is only allowed to be of type `any | unknown`. so rather let devs iterate fast here and pass in the `error` directly then `logger` does some extra work. 

tl;dr: we accept error object, the `trace` here is taken from the `stack` property of `error`

![Screenshot 2024-03-21 at 4 13 44 PM](https://github.com/isomerpages/isomercms-backend/assets/44049504/fffd447e-f3a7-4aad-8811-a93e123653c4)

Reply:
> it's just that in some logs i've observed, the log metadata is grouped by the origin (aws/dd here)

Ah got it, I understand your intent now! It makes sense for providers andplugins to dump their values in namespaces, since that reduces chances of collision with what the app logs. In our case, since we are the top-level app, there's no need for an additional namespace, we could argue that even the `meta` envelop could be considered unnecessary, but I think it's cheap enough to keep 🤔 


> wrt error/err it's there already, as part of meta - type meta = WarnMeta & ErrorMeta & Record<string, unknown>

But that won't work (in the sense that it won't output any useful to the logs).

Error objects are not iterable/serializable and need custom serialization to work. With the current code, you'd get an empty object in the log:

Example:
```javascript
logger.error({meta: {error: new Error('foo')}})
```
```
{
  "level":"ERROR",
  "time":1711067699209,
  "pid":76651,
  "hostname":"Timothee-Groleaus-MacBook-Pro-14-inch-2021-",
  "meta":{
    "error":{}
  }
}
```

Reply:
Oh, on the namespace thing, pino does have a [built-in functionality](https://github.com/pinojs/pino/blob/master/docs/api.md#nestedkey-string) for that btw

Reply:
as discussed, removed the namespace (also thanks for the link!) and opted to stick with a top level `meta` key